<?xml version="1.0"?>
<launch>

    <arg name="odom_instance" default="local" doc="[local, global]"/>
    <arg name="rviz" default="false" />
    <param name="use_sim_time" value="true" />

    <!-- playing rosbag -->
    <node pkg="rosbag" type="play" name="rosbag_play" output="screen"
        args="
            --quiet
            --clock
            --hz 100
            --topics
                /rs_camera/color/image_raw
                /rs_camera/aligned_depth_to_color/image_raw
                /rs_camera/color/camera_info
                /rs_camera/imu
                /lslidar_point_cloud
                /zed_f9p/fix
            --start=15
            --bags
                /media/nikita/2A44A5CE44A59CD7/Users/Nikitos/Documents/облачные\ хранилища/YandexDisk/универ/5\ уровень/тестовый_заезд/rosbag_nami_3.bag
                /media/nikita/2A44A5CE44A59CD7/Users/Nikitos/Documents/облачные\ хранилища/YandexDisk/универ/5\ уровень/тестовый_заезд/rosbag_nami_4.bag
                /media/nikita/2A44A5CE44A59CD7/Users/Nikitos/Documents/облачные\ хранилища/YandexDisk/универ/5\ уровень/тестовый_заезд/rosbag_nami_5.bag
                /media/nikita/2A44A5CE44A59CD7/Users/Nikitos/Documents/облачные\ хранилища/YandexDisk/универ/5\ уровень/тестовый_заезд/rosbag_nami_6.bag
                /media/nikita/2A44A5CE44A59CD7/Users/Nikitos/Documents/облачные\ хранилища/YandexDisk/универ/5\ уровень/тестовый_заезд/rosbag_nami_7.bag
                /media/nikita/2A44A5CE44A59CD7/Users/Nikitos/Documents/облачные\ хранилища/YandexDisk/универ/5\ уровень/тестовый_заезд/rosbag_nami_8.bag
                /media/nikita/2A44A5CE44A59CD7/Users/Nikitos/Documents/облачные\ хранилища/YandexDisk/универ/5\ уровень/тестовый_заезд/rosbag_nami_9.bag
            "/>

    <!-- restoring tf_static -->
    <include file="$(find tigra_software)/launch/debug/restoring_tf_static.launch">
    </include>

    <!-- imu_filter -->
    <!-- <include file="$(find tigra_software)/launch/drivers/imu_filter.launch">
        <arg name="imu/data_raw" value="rs_camera/imu"/>
        <arg name="imu/data_filtered" value="rs_camera/imu/filtered"/>
    </include> -->
    <!-- base_footprint_stabilized -->
    <!-- <node pkg="rtabmap_util" type="imu_to_tf" name="rtabmap_imu_to_tf" output="screen">
        <param name="fixed_frame_id" value="base_footprint_stabilized"/>
        <param name="base_frame_id" value="base_footprint"/>
        <remap from="imu/data" to="rs_camera/imu/filtered"/>
    </node> -->

    <!-- rdbd_odometry -->
    <include file="$(find tigra_vision)/launch/rgbd_odom.launch">
        <arg name="publish_tf" value="false" />
        <arg name="camera_name" default="rs_camera" />
        <arg name="subscribe_rgbd" default="true" />
    </include>

    <!-- slam -->
    <include file="$(find tigra_vision)/launch/rtabmap_slam.launch">
        <arg name="localization" value="false" />
        <arg name="database_path" value="$(find tigra_maps)/3d_maps/new_map.db" />
        <arg name="rtabmap_viz" value="false" />
        <arg name="publish_tf_map" default="false"/>
        <arg name="lidar_name" default="lslidar" />
        <arg name="camera_name" default="rs_camera" />
        <arg name="gps_name" default="zed_f9p"/>
        <arg name="odom_topic" default="visual_odometry"/>
    </include>

    <!-- ekf_local -->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_local_node" clear_params="true" output="screen">
        <rosparam command="load" file="$(find tigra_software)/config/robot_localization/ekf_local_nami.yaml" />
        <param name="publish_tf" value="true" />
        <remap from="odometry/filtered" to="odometry/filtered/local" />
        <remap from="diagnostics" to="ekf_local_diagnostics"/>
    </node>
    <!-- ekf_global -->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_global_node" output="screen">
        <rosparam command="load" file="$(find tigra_software)/config/robot_localization/ekf_global_nami.yaml" />
        <param name="publish_tf" value="true" />
        <!-- inputs -->
        <remap from="set_pose" to="ekf_global/set_pose"/>
        <!-- outputs -->
        <remap from="odometry/filtered" to="odometry/filtered/global"/>
        <remap from="diagnostics" to="ekf_global_diagnostics"/>
    </node>
    <!-- navsat_transform -->
    <node pkg="robot_localization" type="navsat_transform_node" name="navsat_transform_node" output="screen">
        <rosparam command="load" file="$(find tigra_software)/config/robot_localization/navsat_transform_nami.yaml" />
        <!-- inputs -->
        <remap from="odometry/filtered" to="odometry/filtered/global"/>
        <remap from="gps/fix" to="zed_f9p/fix"/>
        <remap from="imu/data" to="rs_camera/imu"/>
        <!-- outputs -->
        <remap from="odometry/gps" to="odometry/gps"/>
        <remap from="gps/filtered" to="gps/filtered"/>
    </node>
    <!-- set initial datum -->
    <node name="set_datum" pkg="tigra_software" type="set_datum.sh" output="screen" />

    <!-- plotting graphs -->
    <group if="$(eval odom_instance == 'local')">
        <node name="rqt_plot" pkg="rqt_plot" type="rqt_plot"
            args="
                /odometry/filtered/local/pose/covariance[0]
                /odometry/filtered/local/pose/covariance[7]
                /odometry/filtered/local/pose/covariance[35]
                /odometry/filtered/local/twist/covariance[0]
                /odometry/filtered/local/twist/covariance[7]
                /odometry/filtered/local/twist/covariance[35]
                "/>
        <node pkg="rqt_multiplot" type="rqt_multiplot" name="rqt_multiplot"
            args="
                --multiplot-config $(find tigra_software)/config/rqt/rqt_multiplot/ekf_local_nami.xml
                --multiplot-run-all
                "/>
    </group>
    <group if="$(eval odom_instance == 'global')">
        <node name="rqt_plot" pkg="rqt_plot" type="rqt_plot"
            args="
                /odometry/filtered/global/pose/covariance[0]
                /odometry/filtered/global/pose/covariance[7]
                /odometry/filtered/global/pose/covariance[35]
                /odometry/filtered/global/twist/covariance[0]
                /odometry/filtered/global/twist/covariance[7]
                /odometry/filtered/global/twist/covariance[35]
                "/>
        <node pkg="rqt_multiplot" type="rqt_multiplot" name="rqt_multiplot"
            args="
                --multiplot-config $(find tigra_software)/config/rqt/rqt_multiplot/ekf_global_nami.xml
                --multiplot-run-all
                "/>
    </group>
    <!-- viev gps data in foxglove -->
    <include file="$(find foxglove_bridge)/launch/foxglove_bridge.launch">
        <arg name="port" value="8765" />
    </include>

    <node if="$(arg rviz)" name="rviz" pkg="rviz" type="rviz" args="-d $(find tigra_software)/config/rqt/rviz/rosbag_nami.rviz" />
    <node name="rqt_console" pkg="rqt_console" type="rqt_console" />

</launch>
